### PR TITLE
Linux: fixed RCInput bug

### PR DESCRIPTION
**Pull request by tridge**
this fixes a bug in the Linux RCInput that was fixed quite a long time ago for HAL_PX4. It bit me today on the Disco when it was entering failsafe while it had good R/C input. The problem was another library calling read() on an input was clearing the new input flag, so the main loop didn't see new input

### CODE REVIEW

File: libraries/AP_HAL/RCInput.h [hunk 1]
```diff
@@ -17,12 +17,9 @@ class AP_HAL::RCInput {
...
 
     /**
-     * Return true if there has been new input since the last read()
-     * call. This call also clears the new_input flag, so once it
-     * returns true it won't return true again until another frame is
-     * received.
+     * Return true if there has been new input since the last call to new_input()
      */
-    virtual bool new_input() = 0;
+    virtual bool new_input(void) = 0;
```
Comment (lucasdemarchi) [inline libraries/AP_HAL/RCInput.h]: you know that in C++ the void in the arg list is not required and often avoided, right?

File: libraries/AP_HAL_Linux/RCInput.h [hunk 2]
```diff
@@ -47,7 +47,8 @@ class RCInput : public AP_HAL::RCInput {
     void _process_rc_pulse(uint16_t width_s0, uint16_t width_s1);
     void _update_periods(uint16_t *periods, uint8_t len);
 
-    volatile bool new_rc_input;
+    volatile uint32_t rc_input_count;
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/RCInput.h]: I don't think this use of volatile is correct.... it doesn't protect against one processor seeing the update to rc_input_count before other fields (e.g.  _num_channels). We should be using an atomic type here to guarantee there's a write fence before. See the atomic usage in RingBuffer.h.

**Comment by lucasdemarchi** [comment]
Applied, but see comments.

File: libraries/AP_HAL/RCInput.h (hunk 1 above)
Comment (tridge) [inline libraries/AP_HAL/RCInput.h]: I find it clearer, it came about as I initially added a parameter, then re-worked without it, replacing with void

**Pull request closed without merging**